File: lib/java/src/main/java/com/workiva/frugal/registry/FRegistry.java
package com.workiva.frugal.registry;

import com.workiva.frugal.FContext;
import org.apache.thrift.TException;

/**
 * Registry is responsible for multiplexing received messages to the appropriate callback.
 */
public interface FRegistry {

    /**
     * Register a callback for the given FContext.
     *
     * @param context the FContext to register.
     * @param callback the callback to register.
     * @throws TException if the given context is already registered to a callback.
     */
    void register(FContext context, FAsyncCallback callback) throws TException;

    /**
     * Unregister the callback for the given FContext.
     *
     * @param context the FContext to unregister.
     */
    void unregister(FContext context);

    /**
     * Dispatch a single Frugal message frame.
     *
     * @param frame an entire Frugal message frame.
     * @throws TException if execution failed.
     */
    void execute(byte[] frame) throws TException;

    /**
     * Interrupt any registered contexts.
     */
    void close();
}

File: lib/src/registry/f_registry.dart
part of frugal;

/// Invoked when a message frame is received. Any error thrown by
/// this callback is considered fatal and will result in the
/// shutdown of the transport.
typedef void FAsyncCallback(TTransport);

/// Responsible for multiplexing received messages to the
/// appropriate callback.
abstract class FRegistry {
  /// Register a callback for the given Context.
  void register(FContext ctx, FAsyncCallback callback);

  /// Unregister a callback for the given Context.
  void unregister(Context);

  /// Dispatch a single Frugal message frame.
  void execute(Uint8List);
}
